pilot/cmd/pilot-agent/readiness.go

```diff
@@ -0,0 +1,239 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+
+	"errors"
+
+	"time"
+
+	"strconv"
+	"strings"
+	"sync"
+
+	"github.com/hashicorp/go-multierror"
+)
+
+const (
+	statsRequestTimeout     = time.Second * 2
+	readyStalenessThreshold = time.Minute * 5
+	statXdsActive           = "cluster.xds-grpc.upstream_cx_active"
+	statCdsUpdates          = "cluster_manager.cds.update_success"
+	statWarmingClusters     = "cluster_manager.warming_clusters"
+	statLdsUpdates          = "listener_manager.lds.update_success"
+	statWarmingListeners    = "listener_manager.total_listeners_warming"
+)
+
+type readinessProbe struct {
+	adminPort         int
+	lastReadySnapshot *statsSnapshot
+	lastErr           error
+	mutex             sync.Mutex
+}
+
+func newReadinessProbe(adminPort int) *readinessProbe {
+	return &readinessProbe{
+		adminPort: adminPort,
+	}
+}
+
+func (p *readinessProbe) checkReady() error {
```

Comment:
I'm completely confused by this code - we get the stats, check them, return a result - all stateless.
Why do we need all the states and mutexes and transitions ? If they are really needed - please add comments explaining.

Reply:
The attempt here is to deal with staleness in the event that we have temporarily lost a connection to Pilot. So we need to maintain some minimal state to determine that we've gotten additional pilot updates after a re-connect.